### PR TITLE
[SDPatternMatch] Add m_CondCode, m_NoneOf, and some SExt improvements

### PR DESCRIPTION
**Pull request by mshockwave**
- Add m_CondCode to match the ISD::CondCode value from CondCodeSDNode
  - Add m_NoneOf combinator
  - m_SExt now recognizes sext_inreg

### CODE REVIEW

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 1]
```diff
@@ -718,6 +723,20 @@ inline SpecificInt_match m_Zero() { return m_SpecificInt(0U); }
 inline SpecificInt_match m_One() { return m_SpecificInt(1U); }
 inline SpecificInt_match m_AllOnes() { return m_SpecificInt(~0U); }
 
+// FIXME: We probably ought to move constant matchers before
+// most of the others so that m_ZExt/m_ZExtOrSelf can be
+// with other extension matchers.
+template <typename Opnd> inline auto m_ZExt(const Opnd &Op) {
+  return m_AnyOf(UnaryOpc_match<Opnd>(ISD::ZERO_EXTEND, Op),
+                 m_And(Op, m_AllOnes()));
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:731]: m_And(Op, m_AllOnes()) is not a zero extend. That's equivalent to `Op`. No bits are cleared.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 2]
```diff
@@ -366,6 +379,10 @@ template <typename... Preds> Or<Preds...> m_AnyOf(Preds &&...preds) {
   return Or<Preds...>(std::forward<Preds>(preds)...);
 }
 
+template <typename... Preds> auto m_NoneOf(Preds &&...preds) {
+  return Not{m_AnyOf(std::forward<Preds>(preds)...)};
```
Comment (marcauberer) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:383]: Maybe we want to have something like `m_not()` to wrap the struct instantiation?

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 1 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: I must be tired and mixed that with something else. It's removed now.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 2 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done. I created a `m_IsNot` to wrap the `Not` struct instantiation.

**Review by marcauberer — APPROVED** [review]

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 3]
```diff
@@ -358,6 +358,23 @@ struct Or<Pred, Preds...> : Or<Preds...> {
...
+
+  template <typename MatchContext>
+  bool match(const MatchContext &Ctx, SDValue N) {
+    return !P.match(Ctx, N);
+  }
+};
+// Explicit deduction guide.
+template <typename Pred> Not(const Pred &P) -> Not<Pred>;
+
+template <typename Pred> inline Not<Pred> m_IsNot(const Pred &P) {
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:374]: Is that would PatternMatch.h calls m_Unless?

Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: That would indeed be more consistent. It's updated now.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 4]
```diff
@@ -758,6 +778,40 @@ inline auto m_False() {
...
 }
 
+struct CondCode_match {
+  std::optional<ISD::CondCode> CCToMatch;
+  ISD::CondCode *BindCC;
+
+  explicit CondCode_match(ISD::CondCode CC) : CCToMatch(CC), BindCC(nullptr) {}
+
+  explicit CondCode_match(ISD::CondCode *CC)
+      : CCToMatch(std::nullopt), BindCC(CC) {}
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:788]: The CCToMatch constructor as unneeded here.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 5]
```diff
@@ -758,6 +778,40 @@ inline auto m_False() {
       m_Value()};
 }
 
+struct CondCode_match {
+  std::optional<ISD::CondCode> CCToMatch;
+  ISD::CondCode *BindCC;
+
+  explicit CondCode_match(ISD::CondCode CC) : CCToMatch(CC), BindCC(nullptr) {}
```
Comment (topperc) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:785]: Uses `ISD::CondCode *BindCC = nullptr` in the class body and you don't have to mention it here.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 4 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Fixed.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h (hunk 5 above)
Comment (mshockwave) [reply llvm/include/llvm/CodeGen/SDPatternMatch.h]: Done.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**